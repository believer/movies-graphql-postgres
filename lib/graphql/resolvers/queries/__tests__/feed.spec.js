const feed = require('../feed')
const { verifyToken } = require('../../../../services/token')

jest.mock('../../../../services/token', () => ({
  verifyToken: jest.fn().mockImplementation(() =>
    Promise.resolve({
      id: 2,
    })
  ),
}))

describe('queries/feed', () => {
  let context

  beforeEach(() => {
    context = {
      db: {
        query: jest.fn().mockReturnValue('dbData'),
      },
      token: 'test',
    }
  })

  afterEach(() => {
    context.db.query.mockClear()
  })

  it('should verify token', () => {
    feed(null, {}, context)

    expect(verifyToken).toHaveBeenCalledWith('test')
  })

  it('should query for a specific user', async () => {
    await feed(null, { limit: 10, offset: 0 }, context)

    expect(context.db.query.mock.calls[0][0]).toMatchSnapshot()
  })

  it('should return database data', async () => {
    const response = await feed(null, { limit: 10, offset: 0 }, context)

    expect(response).toEqual('dbData')
  })

  it('should handle errors', async () => {
    context.db.query.mockImplementation(() => Promise.reject('nope'))

    try {
      await feed(null, { limit: 10, offset: 0 }, context)
    } catch (e) {
      expect(e.message).toMatch(/nope/)
    }
  })
})
