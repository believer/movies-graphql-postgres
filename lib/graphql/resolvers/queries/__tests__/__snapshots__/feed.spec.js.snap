// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries/feed should query for a specific user 1`] = `
"
      SELECT 
        distinct on (m.id, v.view_date) m.*, 
        usr.user,
        
  img.poster, 
  img.backdrop,
  a.actors, 
  d.directors,
  co.composers, 
  w.writers,
  g.genres, 
  c.countries, 
  l.languages, 
  pc.production_companies

      FROM 
        (
          SELECT  * 
          FROM views 
          WHERE user_id = 2
          ORDER BY view_date DESC
          OFFSET 0 
          LIMIT 10
        ) as v, 
        LATERAL (
          SELECT * FROM movies WHERE id = v.movie_id
        ) as m, 
        LATERAL (
          SELECT
            jsonb_build_object(
              'id', id,
              'name', name,
              'email', email) as user
          FROM users WHERE id = v.user_id
        ) as usr,
        
  LATERAL (
    SELECT * FROM images WHERE movie_id = m.id
  ) as img,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name)
          FROM 
            movie_genres AS comb 
            INNER JOIN genres AS p ON p.id = comb.genre_id
          WHERE 
            comb.movie_id = m.id
        ) as genres
    ) as g 
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name)
          FROM 
            movie_countries AS comb 
            INNER JOIN countries AS p ON p.id = comb.country_id
          WHERE 
            comb.movie_id = m.id
        ) as countries
    ) as c 
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name)
          FROM 
            movie_languages AS comb 
            INNER JOIN languages AS p ON p.id = comb.language_id
          WHERE 
            comb.movie_id = m.id
        ) as languages
    ) as l 
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name)
          FROM 
            movie_production_companies AS comb 
            INNER JOIN production_companies AS p ON p.id = comb.production_company_id
          WHERE 
            comb.movie_id = m.id
        ) as production_companies
    ) as pc 
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name) 
          FROM 
            cast_and_crew AS cac 
            INNER JOIN people AS p ON p.id = cac.person_id 
          WHERE 
            cac.movie_id = m.id 
            AND cac.role_id = 1
        ) as directors
    ) as d
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name) 
          FROM 
            cast_and_crew AS cac 
            INNER JOIN people AS p ON p.id = cac.person_id 
          WHERE 
            cac.movie_id = m.id 
            AND cac.role_id = 2
        ) as actors
    ) as a
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name) 
          FROM 
            cast_and_crew AS cac 
            INNER JOIN people AS p ON p.id = cac.person_id 
          WHERE 
            cac.movie_id = m.id 
            AND cac.role_id = 3
        ) as writers
    ) as w
  ,
  
    LATERAL (
      SELECT 
        ARRAY (
          SELECT 
            trim(p.name) 
          FROM 
            cast_and_crew AS cac 
            INNER JOIN people AS p ON p.id = cac.person_id 
          WHERE 
            cac.movie_id = m.id 
            AND cac.role_id = 4
        ) as composers
    ) as co
  

      ORDER BY 
        v.view_date DESC
    "
`;
